Preserve image attributes when cropping a custom logo image

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

This PR aims to implement preserving attributes when cropping a custom logo image.

Steps to test this PR: https://core.trac.wordpress.org/ticket/37750#comment:26

Trac ticket: https://core.trac.wordpress.org/ticket/37750

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Copy attachment's description, title, and caption over to the cropped image (if they are set).

anton-vlasenko committed:
Improve detection of empty strings.

anton-vlasenko committed:
1. Rename $attachmenet to $original_attachement.
2. Copy image alt attribute as well.

anton-vlasenko committed:
Don't call get_post_meta to retrieve _wp_attachment_image_alt option.

anton-vlasenko committed:
1. Don't call wp_basename if the image has custom title.
2. Add comments.

anton-vlasenko committed:
Add test template.

anton-vlasenko committed:
Implement test.
This is WIP.

anton-vlasenko committed:
1. Fix code style.
2. Fix the test. Delete cropeed image.

anton-vlasenko committed:
Fix the test: properly delete cropped images.

anton-vlasenko committed:
Remove unused variables.

anton-vlasenko committed:
Copy the post title attribute from original image.

anton-vlasenko committed:
Fix covers tags.

anton-vlasenko committed:
Fix PHPDOC description.

dream-encode commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3975,18 +3975,34 @@ function wp_ajax_crop_image() {
 			$image_type = ( $size ) ? $size['mime'] : 'image/jpeg';
+			/** @var WP_Post $original_attachment */
+			$original_attachment = get_post( $attachment_id );
+			$has_title           = 0 < mb_strlen( trim( $original_attachment->post_title ) );
```
```suggestion
			$has_title           = mb_strlen( trim( $original_attachment->post_title ) ) > 0;
```
A Yoda condition here degrades readability, per the WPCS(https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/)

anton-vlasenko replied:
Thank you for the feedback! Fixed in fdaa748a589a8cb0ab7e21b624e38f35284027df

dream-encode commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3975,18 +3975,34 @@ function wp_ajax_crop_image() {
+			/** @var WP_Post $original_attachment */
+			$original_attachment = get_post( $attachment_id );
+			$has_title           = 0 < mb_strlen( trim( $original_attachment->post_title ) );
+			$has_description     = 0 < mb_strlen( trim( $original_attachment->post_content ) );
```
```suggestion
			$has_description     = mb_strlen( trim( $original_attachment->post_content ) ) > 0;
```
A Yoda condition here degrades readability, per the WPCS(https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/)

anton-vlasenko replied:
Fixed in https://github.com/WordPress/wordpress-develop/commit/fdaa748a589a8cb0ab7e21b624e38f35284027df

dream-encode commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,150 @@
+		$this->assertInstanceOf( WP_Post::class, $cropped_attachment );
+		$this->assertNotEmpty( $attachment->post_title );
+		$this->assertNotEmpty( $cropped_attachment->post_title );
+		$this->assertSame( $attachment->post_title, $cropped_attachment->post_title );
```
There's a few assertions here in the same test.  In this instance, it's best to include the `$message` parameter when you have multiple assertions in a test.  See https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions

anton-vlasenko replied:
I agree. Thank you for the feedback.
Fixed in f96f9d31420ce77768c94bddfeb9babedafaa716

dream-encode commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,150 @@
+class Tests_Ajax_CropImage extends WP_Ajax_UnitTestCase {
+
+	/**
+	 * @covers wp_ajax_crop_image
```
Should probably note `@ticket 37750` here as well.

anton-vlasenko replied:
Thank you! Fixed in 4a548660707894c1983fae48939ce1af07d98d78

dream-encode commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,150 @@
+class Tests_Ajax_CropImage extends WP_Ajax_UnitTestCase {
+
+	/**
+	 * @covers wp_ajax_crop_image
```
```suggestion
	 * @covers ::wp_ajax_crop_image()
```

anton-vlasenko replied:
Thank you for the feedback.
Fixed in 4a548660707894c1983fae48939ce1af07d98d78.
I wouldn't use parenthesis at the end because it seems they are not needed according to the PHPUnit documentation.

hellofromtonya replied:
Yes, I agree with not including the `()` at the end of the `@covers` annotation.  https://github.com/WordPress/wordpress-develop/commit/4a548660707894c1983fae48939ce1af07d98d78 LGTM 👍

dream-encode commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	/**
+	 * @covers wp_ajax_crop_image
```
```suggestion
	 * @covers ::wp_ajax_crop_image()
```

anton-vlasenko replied:
Fixed in 4a548660707894c1983fae48939ce1af07d98d78

dream-encode reviewed (commented):
More to address here in the actual logic, but these are the things that jump out at me straight away.

anton-vlasenko committed:
Don't use Yoda conditions for comparisons.

anton-vlasenko committed:
Update PHPDOC block.

anton-vlasenko committed:
Add fail messages to assertions.

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3975,18 +3975,34 @@ function wp_ajax_crop_image() {
+			/** @var WP_Post $original_attachment */
+			$original_attachment = get_post( $attachment_id );
+			$has_title           = mb_strlen( trim( $original_attachment->post_title ) ) > 0;
+			$has_description     = mb_strlen( trim( $original_attachment->post_content ) ) > 0;
```
A simplification (see it in action https://3v4l.org/hCPCX):

```php
( '' !== trim( $original_attachment->post_title ) );
```

```suggestion
			$has_title           = ( '' !== trim( $original_attachment->post_title ) );
			$has_description     = ( '' !== trim( $original_attachment->post_content ) );
```

anton-vlasenko replied:
I absolutely agree. It's more efficient to compare values with an empty string.
Fixed in a2e8c8690e82677dfeaea998b027b10f72e6cbdd

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3975,18 +3975,34 @@ function wp_ajax_crop_image() {
 			);
 
+			// Copy the image caption attribute (post_excerpt field) from the original image.
+			if ( mb_strlen( trim( $original_attachment->post_excerpt ) ) ) {
```
Same here
```suggestion
			if ( '' !== trim( $original_attachment->post_excerpt ) ) {
```

anton-vlasenko replied:
Fixed in https://github.com/WordPress/wordpress-develop/commit/a2e8c8690e82677dfeaea998b027b10f72e6cbdd

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3975,18 +3975,34 @@ function wp_ajax_crop_image() {
 			$metadata      = wp_generate_attachment_metadata( $attachment_id, $cropped );
 
+			// Copy the image alt text attribute from the original image.
+			if ( mb_strlen( trim( $original_attachment->_wp_attachment_image_alt ) ) ) {
```
```suggestion
			if ( '' !== trim( $original_attachment->_wp_attachment_image_alt ) ) {
```

anton-vlasenko replied:
Fixed in https://github.com/WordPress/wordpress-develop/commit/a2e8c8690e82677dfeaea998b027b10f72e6cbdd

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+/**
+ * Class for testing ajax crop image functionality.
+ *
+ * @group ajax
```
As this test class is specific for one function, the `@covers` annotation can be set at the class level rather than at the test method level.

```suggestion
 * @group ajax
 * @covers ::wp_ajax_crop_image
```

anton-vlasenko replied:
Good to know. Thank you.
Fixed in 412e86234bf99b738912fecd5d84be4897a4eb30

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+	 */
+	public function test_it_copies_metadata_from_original_image() {
+
+		// Become an administrator.
+		$this->_setRole( 'administrator' );
```
As both tests use this, I'd suggest moving it to `set_up()` fixture method instead as then the tests are setup and ready to go.

anton-vlasenko replied:
Fixed in 412e86234bf99b738912fecd5d84be4897a4eb30

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+
+	/**
+	 * @param array $response Response to validate.
+	 *
+	 * @return void
```
```suggestion
```

Core doesn't use `@return void`.

anton-vlasenko replied:
Fixed in a4cb089d043fefcc7e957e13428f4e0da059ffbd

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+	 * Prepares $_POST for crop-image ajax action.
+	 *
+	 * @param WP_Post $attachment
+	 *
+	 * @return void
```
```suggestion
```

Core doesn't use `@return void`.

anton-vlasenko replied:
Fixed in a4cb089d043fefcc7e957e13428f4e0da059ffbd

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+	 *
+	 * @return WP_Post
+	 */
+	private function create_attachment( $with_metadata = true ) {
+		$uniq_id = uniqid( 'crop-image-ajax-action-test-' );
+		$object  = array(
+			'post_title'     => $with_metadata ? 'Title ' . $uniq_id : '',
+			'post_content'   => $with_metadata ? 'Description ' . $uniq_id : '',
+			'post_mime_type' => 'image/jpg',
+			'guid'           => 'http://localhost/foo.jpg',
+			'context'        => 'custom-logo',
+			'post_excerpt'   => $with_metadata ? 'Caption ' . $uniq_id : '',
+		);
+
+		$test_file        = DIR_TESTDATA . '/images/test-image.jpg';
+		$upload_directory = wp_upload_dir();
+		$uploaded_file    = $upload_directory['path'] . '/' . $uniq_id . '.jpg';
+		$filesystem       = new WP_Filesystem_Direct( true );
+		$filesystem->copy( $test_file, $uploaded_file );
+		$attachment_id = wp_insert_attachment( $object, $uploaded_file );
+
+		if ( $with_metadata ) {
+			update_post_meta( $attachment_id, '_wp_attachment_image_alt', wp_slash( 'Alt ' . $uniq_id ) );
+		}
+
+		return get_post( $attachment_id );
+	}
```
To avoid duplication and inconsistencies, I'd suggest instead to use `$this->_make_attachment()` helper method (which is available through inheritance coming from `WP_UnitTestCase_Base`). If `$with_metadata` is `true`, then update the post and its metadata. Doing this ensures the logic needed to create an attachment is centralized in one location.

anton-vlasenko replied:
> I'd suggest instead to use $this->_make_attachment() helper method 

That's good to know. Fixed in a39a55141d5da872225890a2054722f567a657de

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+ *
+ * @group ajax
+ */
+class Tests_Ajax_CropImage extends WP_Ajax_UnitTestCase {
```
The class name should follow this convention `Tests_[group]_FunctionName`, which changes this test class to `Tests_Ajax_WpAjaxCropImage`.

and the file should be camelCase of the function name, e.g. `wpAjaxCropImage.php`.

anton-vlasenko replied:
That's good to know! Thank you.
Fixed in 743688eec4566886bdfac2f9c0312e224eb19bc0
P.S. It seems that all classes/files in the `tests/phpunit/tests/ajax` folder have to be renamed because they don't follow this convention.

hellofromtonya replied:
>P.S. It seems that all classes/files in the `tests/phpunit/tests/ajax` folder have to be renamed because they don't follow this convention.

This is true throughout the test suite. Can be remedied as part of https://core.trac.wordpress.org/ticket/54725. Feel free to create a PR.

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+		$this->assertSame( $attachment->post_excerpt, $cropped_attachment->post_excerpt, 'post_excerpt value should be copied over to the cropped attachment' );
+		$this->assertSame( $attachment->_wp_attachment_image_alt, $cropped_attachment->_wp_attachment_image_alt, '_wp_attachment_image_alt value should be copied over to the cropped attachment' );
+
+		wp_delete_attachment( $attachment->ID, true );
+		wp_delete_attachment( $cropped_attachment->ID, true );
```
This logic will not be reached if a test fails. Instead move it to `tear_down()` which means the ID of each will need to become a property for `tear_down()` to have access.

anton-vlasenko replied:
I've already tried to move this code to `tear_down`, and the attached images weren't deleted. I've just realised (thanks to your comment) that I can fix it by calling `parent::tear_down()` at the very end of my `tear_down` method. 🤦

Otherwise, DB entities will be rolled back in the parent `tear_down` method, and the `wp_delete_attachment` method will not work.
Fixed in c55674c7b8e71c6043f2bd6aa35cf42d5c6d51be

hellofromtonya commented on tests/phpunit/tests/ajax/CropImage.php:
```diff
@@ -0,0 +1,158 @@
+		$this->assertEmpty( $cropped_attachment->post_excerpt, 'post_excerpt value must be empty if it\'s empty in the original attachment' );
+		$this->assertEmpty( $cropped_attachment->_wp_attachment_image_alt, '_wp_attachment_image_alt value must be empty if it\'s empty in the original attachment' );
+
+		wp_delete_attachment( $attachment->ID, true );
+		wp_delete_attachment( $cropped_attachment->ID, true );
```
same note as previous test.

anton-vlasenko replied:
Fixed in https://github.com/WordPress/wordpress-develop/commit/c55674c7b8e71c6043f2bd6aa35cf42d5c6d51be

hellofromtonya reviewed (changes requested):
Overall looks good. Some changes for consistency.

anton-vlasenko committed:
Second attempt to implement tear_down method.

anton-vlasenko committed:
Refactoring: move $this->_setRole( 'administrator' ); to the set_up method.

anton-vlasenko committed:
Core doesn't use @return void tags.

anton-vlasenko committed:
Refactor @covers tag.

anton-vlasenko committed:
The class name should follow this convention Tests_[group]_FunctionName, which changes this test class to Tests_Ajax_WpAjaxCropImage.
And the file should be camelCase of the function name, e.g. wpAjaxCropImage.php.

anton-vlasenko committed:
Don't use mb_strlen to detect empty strings.

anton-vlasenko committed:
Refactor Tests_Ajax_WpAjaxCropImage::make_attachment and use the _make_attachment method to create a test post.

hellofromtonya committed:
Move fixture methods above tests for consistency.

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3975,18 +3975,37 @@ function wp_ajax_crop_image() {
 				'context'        => $context,
 			);
 
+			// Copy the image description attribute (post_content field) from the original image.
+			if ( '' !== trim( $original_attachment->post_content ) ) {
+				$object['post_content'] = $original_attachment->post_content;
+			}
```
Instead of assigning the `post_content` above and then overwriting it, move this inline into the array above.

hellofromtonya replied:
Currently updating and will push shortly

hellofromtonya committed:
Check if original title was edited

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3970,23 +3970,43 @@ function wp_ajax_crop_image() {
+			$use_original_title       = (
+				'' !== trim( $original_attachment->post_title ) &&
+				// Check if the original image's title was edited.
+				pathinfo( $parent_basename, PATHINFO_FILENAME ) !== $original_attachment->post_title
```
@anton-vlasenko I added this check to only use the original image's title _if it was edited_. When there's no title in the image's metadata, by default the image's filename is set as the `post_title`. This check is making sure the `post_title` is not the same as that default behavior. If no, then it will use the original image's title.

hellofromtonya replied:
The tests will need updating though.

anton-vlasenko replied:
Thank you!
I'm working on updating the tests.

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3970,23 +3970,43 @@ function wp_ajax_crop_image() {
+				// Check if the original image's title was edited.
+				pathinfo( $parent_basename, PATHINFO_FILENAME ) !== $original_attachment->post_title
+			);
+			$use_original_description = ( '' !== trim( $original_attachment->post_content ) );
```
@anton-vlasenko I moved this back here for consistency and then renamed to make it more clear _what_ the value is.

anton-vlasenko replied:
Thank you, @hellofromtonya.

anton-vlasenko committed:
Fixed incorrect detection of an "edited" state.
We have to call sanitize_file_name to escape $original_attachment->post_title, otherwise this comparison will not work as intended.
E.g. post_title can have spaces, but they get replaced with dashes in the filename.
Also, sometimes post_title contains file extension, sometimes it doesn't.

anton-vlasenko committed:
Improve detection of the "edited" state.
We have to call sanitize_file_name to escape $original_attachment->post_title, otherwise this comparison will not work as intended.
E.g. post_title can have spaces, but they get replaced with dashes in the filename.

anton-vlasenko committed:
Add assertions for cases when post_title gets populated.

anton-vlasenko committed:
Fix the code style.

anton-vlasenko committed:
Fix code style.

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3970,23 +3970,45 @@ function wp_ajax_crop_image() {
+			$sanitized_post_title     = sanitize_file_name( $original_attachment->post_title );
+			$use_original_title       = (
+				'' !== trim( $original_attachment->post_title ) &&
+				// Check if the original image's title was edited.
```
```suggestion
				/*
				 * Check if the original image has a title other than the "filename" default,
				 * meaning the image had a title when originally uploaded or its title was edited.
				 */
```

hellofromtonya committed:
Improve image title check comment.

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3970,23 +3970,48 @@ function wp_ajax_crop_image() {
+			$original_attachment      = get_post( $attachment_id );
+			$sanitized_post_title     = sanitize_file_name( $original_attachment->post_title );
+			$use_original_title       = (
+				'' !== trim( $original_attachment->post_title ) &&
```
```suggestion
				( '' !== trim( $original_attachment->post_title ) ) &&
```
for consistency with the other checks.

hellofromtonya committed:
Wrap conditional for consistency.

hellofromtonya commented on tests/phpunit/tests/ajax/wpAjaxCropImage.php:
```diff
@@ -0,0 +1,223 @@
+ * Admin Ajax functions to be tested.
+ */
+require_once ABSPATH . 'wp-admin/includes/ajax-actions.php';
+
```
```suggestion
```

hellofromtonya committed:
Formatting - remove empty line.

hellofromtonya reviewed (approved):
The implementation and tests LGTM 👍
Approving - pending input from @felixarntz ([see context in the Trac ticket comment](https://core.trac.wordpress.org/ticket/37750#comment:29)).

felixarntz commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3970,23 +3970,48 @@ function wp_ajax_crop_image() {
 				'context'        => $context,
 			);
 
+			// Copy the image caption attribute (post_excerpt field) from the original image.
+			if ( '' !== trim( $original_attachment->post_excerpt ) ) {
+				$object['post_excerpt'] = $original_attachment->post_excerpt;
+			}
```
Bit of a nit-pick, but I think it would make sense to include this immediately in the associative array above, right after `post_title` and `post_content`, which would clarify the relationship of the three when reading the code. The default for this would need to be the empty string (`''`), which is the default for `wp_insert_post` (and thus `wp_insert_attachment`).

anton-vlasenko replied:
Thank you for the feedback.
I intentionally didn't include it in the associative array.
Why? I prefer not to specify a default value if the task does not require that.
I prefer to let `wp_insert_attachment` decide what value to use as default.

felixarntz replied:
That's a fair point. Works for me 👍

felixarntz commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -3970,23 +3970,48 @@ function wp_ajax_crop_image() {
 			$attachment_id = wp_insert_attachment( $object, $cropped );
 			$metadata      = wp_generate_attachment_metadata( $attachment_id, $cropped );
 
+			// Copy the image alt text attribute from the original image.
+			if ( '' !== trim( $original_attachment->_wp_attachment_image_alt ) ) {
+				update_post_meta( $attachment_id, '_wp_attachment_image_alt', wp_slash( $original_attachment->_wp_attachment_image_alt ) );
+			}
```
Similarly to my above comment, we could integrate this into the associative array passed to `wp_insert_attachment` above, which would colocate it with the other other related data. Other than in my previous comment though, here it would be more about avoiding an extra call to `update_post_meta`.

We can use the `meta_input` key (supported by `wp_insert_post` and thus also by `wp_insert_attachment`) to pass this information, basically like `$object['meta_input'] = array( '_wp_attachment_image_alt' => wp_slash( $original_attachment->_wp_attachment_image_alt ) )`.

anton-vlasenko replied:
I absolutely agree. I didn't know `wp_insert_post` supports the `meta_input` key.
Fixed in 9e2a262b7e01d26600d62375f2bd16536e7e931e
cc @hellofromtonya 
I've implemented the change suggested in @felixarntz`s [comment](https://github.com/WordPress/wordpress-develop/pull/2441#discussion_r834770715) above. I tested it manually and it worked as intended. Unit tests also passed.
But if that change will require retesting all use cases manually, please let me know, and I will revert that change.
I also don't want to block this PR from getting merged.

felixarntz reviewed (approved):
This looks great @anton-vlasenko! I left two very minor comments for your consideration, although they certainly don't need to block this from getting merged. Please take a look.

@hellofromtonya I've left a comment in https://core.trac.wordpress.org/ticket/37750#comment:33 as well.

anton-vlasenko committed:
Don't call update_post_meta because the _wp_attachment_image_alt attribute can be set with wp_insert_attachment.

hellofromtonya commented:
Committed via changeset https://core.trac.wordpress.org/changeset/53027.
